tensorboard/backend/event_processing/db_import_multiplexer.py

```diff
@@ -92,97 +106,230 @@ def AddRunsFromDirectory(self, path, name=None):
+    """
+    self._db_path = db_path
+    self._writer_fn_cache = {}
+
+  def _get_writer_fn(self, event_batch):
+    key = (event_batch.experiment_name, event_batch.run_name)
+    if key in self._writer_fn_cache:
+      return self._writer_fn_cache[key]
     with tf.Graph().as_default():
-      self._placeholder = tf.placeholder(shape=[], dtype=tf.string)
+      placeholder = tf.placeholder(shape=[], dtype=tf.string)
       writer = tf.contrib.summary.create_db_writer(
-          db_path, experiment_name=experiment_name, run_name=run_name)
+          self._db_path,
+          experiment_name=event_batch.experiment_name,
+          run_name=event_batch.run_name)
       with writer.as_default():
         # TODO(nickfelt): running import_event() one record at a time is very
         #   slow; we should add an op that accepts a vector of records.
-        self._import_op = tf.contrib.summary.import_event(self._placeholder)
-      self._session = tf.Session()
-      self._session.run(writer.init())
-
-  def Import(self):
-    """Imports all events added since the last call to `Import`."""
-    for record in self._directory_generator.Load():
-      self._session.run(self._import_op, feed_dict={self._placeholder: record})
+        import_op = tf.contrib.summary.import_event(placeholder)
+      session = tf.Session()
+      session.run(writer.init())
+      def writer_fn(event_proto):
+        session.run(import_op, feed_dict={placeholder: event_proto})
+    self._writer_fn_cache[key] = writer_fn
+    return writer_fn
+
+  def write_batch(self, event_batch):
+    start = time.time()
+    writer_fn = self._get_writer_fn(event_batch)
+    for event_proto in event_batch.events:
+      writer_fn(event_proto)
+    elapsed = time.time() - start
+    tf.logging.debug(
+        'ImportOpEventSink.WriteBatch() took %0.3f sec for %s events', elapsed,
+        len(event_batch.events))
+
+
+class _SqliteWriterEventSink(_EventSink):
+  """Implementation of EventSink using SqliteWriter."""
+
+  def __init__(self, db_connection_provider):
+    """Constructs a SqliteWriterEventSink.
+
+    Args:
+      db_connection_provider: Provider function for creating a DB connection.
+    """
+    self._writer = sqlite_writer.SqliteWriter(db_connection_provider)
+
+  def write_batch(self, event_batch):
+    start = time.time()
+    tagged_data = {}
+    for event_proto in event_batch.events:
+      event = tf.Event.FromString(event_proto)
+      self._process_event(event, tagged_data)
+    if tagged_data:
+      self._writer.write_summaries(
+          tagged_data,
+          experiment_name=event_batch.experiment_name,
+          run_name=event_batch.run_name)
+    elapsed = time.time() - start
+    tf.logging.debug(
+        'SqliteWriterEventSink.WriteBatch() took %0.3f sec for %s events',
+        elapsed, len(event_batch.events))
+
+  def _process_event(self, event, tagged_data):
+    """Processes a single tf.Event and records it in tagged_data."""
```

Comment:
If we want to support more types of events, we would expand this method, right? This file is already quite dense so I think we will _eventually_ (no need to do it in this PR) have to factor this class out to a separate file :(

Reply:
As indicated by the TODOs below, yeah I foresee expanding this method/class somewhat.  I don't think it would get too large though, since much of the actual logic to implement e.g. graph writing or purging would live in SqliteWriter.

Given that this file is only a few hundred lines, I'm not sure I see a pressing need to factor this class out.  Doing so is a bit of a pain because if we want to avoid a circular dependency, it also means factoring out EventSink.